daemon/cmd/kube_proxy_replacement.go

```diff
@@ -301,6 +301,24 @@ func detectNativeDevices(strict bool) {
 			l.Info("Using auto-derived devices for BPF node port")
 		}
 	}
```

Comment:
```suggestion
	} else if option.Config.EnableNodePort { // both --devices and --direct-routing-device are specified by user
	        // Check whether the DirectRoutingDevice (if specified) is
		// defined within devices and if not, add it.
		if option.Config.DirectRoutingDevice != "" {
			directDev := option.Config.DirectRoutingDevice
			directDevFound := false
			for _, iface := range option.Config.Devices {
				if iface == directDev {
					directDevFound = true
					break
				}
			}
			if !directDevFound {
				option.Config.Devices = append(option.Config.Devices, directDev)
			}
		}
	}
	
```

Reply:
So there is no case where a user will specify a `--direct-routing-device` that will not be auto-detected by the device detection?

Reply:
The proposed else if branch handles the case when user have provided both `--direct-routing-device` and `--devices`, i.e. the case when no auto detection takes place.

Reply:
> The proposed else if branch handles the case when user have provided both --direct-routing-device and --devices, i.e. the case when no auto detection takes place.

Indeed. 

I'm asking, however, about the case where the user specifies `--direct-routing-device` but not `--devices`.

Specifically, I'm asking whether there is a possibility that a panic can happen in this case. That is, can the user define a valid device with `--direct-routing-device` that will not be on the auto-detected list? AFAIU this will cause a panic (see below) with your else if suggestion but not with the original code.



 
Details

* We are indexing `nodePortIPv4Addrs` with `directRoutingIface` in https://github.com/cilium/cilium/blob/89c0f08cc13125deb2a4fe409d465b48301ba34c/pkg/datapath/linux/config/config.go#L341 where `directRoutingIface` is `option.Config.DirectRoutingDevice`. We seem to checking that `directRoutingDevice` is a real iface, since we are getting is index.

* `nodePortIPv4Addrs` is the return of `GetNodePortIPv4AddrsWithDevices` which seems to be copying `ipv4NodePortAddrs`  which is set in `InitNodePortAddrs` https://github.com/cilium/cilium/blob/89c0f08cc13125deb2a4fe409d465b48301ba34c/pkg/node/address.go#L131
* `InitNodePortAddrs` is called by `finishKubeProxyReplacementInit` with `option.Config.Devices` https://github.com/cilium/cilium/blob/89c0f08cc13125deb2a4fe409d465b48301ba34c/daemon/cmd/kube_proxy_replacement.go#L334

So, AFAIU, if we somehow manage to get a valid device in `option.Config.DirectRoutingDevice` that is not in `option.Config.Devices` the agent will panic.

Reply:
Thanks for the detailed comment. I've just tried running with `--direct-routing-devices=$SOME_DEV_WHICH_CANNOT_BE_DETECTED` and without specifying `--devices` - it worked!

Reply:
OK, I guess that's good enough. pushed with the suggested change.